dask_geopandas/hilbert_distance.py

```diff
@@ -0,0 +1,261 @@
+import geopandas
+import numpy as np
+
+# from numba import jit # optional - do we want numba as a dependency for dask-geopandas?
+# ngjit = jit(nopython=True, nogil=True)
+
+
+def _hilbert_distance(gdf, total_bounds, p):
+
+    """
+    Calculate the hilbert distance for a GeoDataFrame based on the mid-point of
+    the bounds for each geom and total bounds of the collection of geoms
+
+    Based on: https://github.com/holoviz/spatialpandas/blob/
+    9252a7aba5f8bc7a435fffa2c31018af8d92942c/spatialpandas/dask.py#L172
+
+    Parameters
+    ----------
+    gdf          : GeoDataFrame
+
+    total_bounds : Total bounds of GeoDataFrame
+
+    p            : Hilbert curve parameter
+
+    Returns
+    ---------
+    Array of hilbert distances for each geom
+    """
+
+    if total_bounds is None:
+        total_bounds = gdf.total_bounds
+
+    # Calculate bounds of each geom
+    bounds = gdf.bounds.to_numpy()
+
+    # Hilbert Side len
+    side_length = 2 ** p
+
+    # Calculate x and y range of total bound coords - returns array
+    geom_ranges = [
+        (total_bounds[0], total_bounds[2]),
+        (total_bounds[1], total_bounds[3]),
+    ]
+    # Calculate mid points for x and y bound coords - returns array
+    geom_mids = [
+        ((bounds[:, 0] + bounds[:, 2]) / 2.0),
+        ((bounds[:, 1] + bounds[:, 3]) / 2.0),
+    ]
+
+    # Empty coord array
+    coords = np.zeros((bounds.shape[0], 2), dtype=np.int64)
+    # Transform continuous int to discrete int for each dimension
+    coords[:, 0] = _continuous_int_to_discrete_int(
+        geom_mids[0], geom_ranges[0], side_length
+    )
+    coords[:, 1] = _continuous_int_to_discrete_int(
+        geom_mids[1], geom_ranges[1], side_length
+    )
+
+    # Calculate hilbert distance
+    hilbert_distances = _distances_from_coordinates(p, coords)
+
+    return hilbert_distances
+
+
+# @ngjit
+def _continuous_int_to_discrete_int(vals, val_range, n):
+
+    """
+    Convert an array of values from continuous data coordinates to discrete
+    int coordinates
+
+    Based on: https://github.com/holoviz/spatialpandas/blob/
+    9252a7aba5f8bc7a435fffa2c31018af8d92942c/spatialpandas/utils.py#L9
+
+    Parameters
+    ----------
+    vals      : Array of continuous coordinates to be
+                ([([val_1, val_2,..., val_n]), array([val_1, val_2,..., val_n])])
+
+    val_range : Ranges of x and y values ([(xmin, xmax), (ymin, ymax)])
+
+    n         : Number of discrete coords (int)
+
+    Returns
+    ---------
+    Array of discrete int coords
+    """
+
+    x_width = val_range[1] - val_range[0]
+    res = ((vals - val_range[0]) * (n / x_width)).astype(np.int64)
```

Comment:
```suggestion
    res = ((vals - val_range[0]) * (n / width)).clip(min=0, max=n).astype(np.int64)
```
(uses ndarray `clip` instead of manual clipping below.

Note: the values will always be unsigned, which implies that you could use `np.uint32` by default - but it is based on the value of n; if `n` <= 65535, you could use `uint16`, etc.

Reply:
> (uses ndarray `clip` instead of manual clipping below.

`clip` is much cleaner but is causing problems right now [because it's not supported by numba](https://github.com/numba/numba/issues/3465#issuecomment-844893052)

> Note: the values will always be unsigned, which implies that you could use `np.uint32` by default - but it is based on the value of n; if `n` <= 65535, you could use `uint16`, etc.

I will leave this unresolved for now.

Reply:
It is OK to clip manually, I didn't realize that `clip` wasn't yet supported in numba (was testing manually without).  This could be added as a follow-up issue that requires the latest version of numba, but seems pretty minor priority-wise.

However, the data type part of the comment may be work further exploration; it may be a way of reducing memory consumption.

Reply:
Data type optimization should move to a separate issue / PR so that you don't need to include it here.